protocol/smart-batching.md

```diff
@@ -0,0 +1,78 @@
+batches. The derivation pipeline is modified to ignore batches where the contract reverts. This
+expands the design space for OP stack rollups running the canonical derivation pipeline.
+
+Example uses:
+
+- TEE-sequencer: L1 enforces that the rollup sequencer runs in a TEE, which enforces additional sequencing policies:
+    - Unichain wants to use a TEE for Flashblocks and verifiable priority ordering. Currently the derivation pipeline cannot check the TEE attestation, so the current implementation is no better than a trusted sequencer. This proposed change would provide an avenue to fix that. It will also be needed for the UVN.
+    - The Celo sequencer will run a TEE that first posts its transaction batches to a fast BFT consensus (run by Espresso Network) and only posts batches to the L1 that were already confirmed by this BFT consensus. This allows users and protocols (exchanges, bridges, etc) to obtain rollup soft confirmations (pre-L1 finality) from the fast BFT consensus that are more secure than confirmations coming from a regular centralized sequencer. Other OP stack chains are interested in adding this protective layer as well.
+- L1 directly verifies that the sequenced batches have been confirmed by an external layer (e.g. Espresso) instead of relying on a TEE.
+- Enabling various decentralized sequencing policies, by deploying an inbox contract with logic for determining which sequencer is eligible to post any given batch (similar to [RFP](https://github.com/ethereum-optimism/ecosystem-contributions/issues/63) that Espresso completed for OP Labs)
+
+## Problem Statement + Context
+
+It is desirable for many chains to customize the behavior of the OP stack. This is always possible
+by forking and modifying the derivation pipeline, but it is desirable to be able to run the canonical stack, with custom behavior as plugins. OP stack natively supports customization of certain aspects of the rollup (such as alt-DA layers) but lacks sufficient customizability to deploy new batching and sequencing policies.
+
+A simple way for rollups to deploy custom batch posting rules without needing to run a fork of
+validators and RPC nodes is to deploy a *batch inbox* contract on L1, which acts as a pre-filter to
+the derivation pipeline, rejecting any batches that don't satisfy the custom rules while
+allowing the derivation pipeline to ingest the remaining batches and execute them as normal.
+
+Unfortunately, this is not quite possible with the current derivation pipeline. While the inbox
+address can be configured to the address of a contract, the contract cannot stop batches from being ingested by the derivation pipeline, because **the derivation pipeline does not check revert status of transactions**. Thus, even if the contract reverts, the reverting transactions can still end up included in an L1 block, and the derivation pipeline will still parse and execute their calldata or
+blob data.
+
+## Proposed Solution
```

Comment:
Right now its not possible to reconfigure the batch inbox given that existing chains send to an address with no know private key generally. Any thoughts on how to handle this?
- Enable ability to modify batch inbox address dynamically
- Only allow new chains to use this feature?

Reply:
Hmm so your saying it's not currently possible to do a chain upgrade to change the inbox address? If not we would definitely want to add this ability

Reply:
It is not currently possible for a chain to change its batch inbox. It is hardcoded as part of the chains genesis config

Reply:
Gotcha, I'll work on adding this, thanks

Reply:
@tynes it looks like the inbox address is currently read from the system config contract, so we propose simply adding an owner-only function to set this address in the contract. Does that work? [2139117](https://github.com/ethereum-optimism/design-docs/pull/229/commits/2139117c3bff4078ce79bfbe2b450a05393edb06)

Reply:
https://github.com/ethereum-optimism/design-docs/pull/229#discussion_r2042870057